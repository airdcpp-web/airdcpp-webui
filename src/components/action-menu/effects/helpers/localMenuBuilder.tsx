import { toActionI18nKey } from '@/utils/ActionUtils';

import * as UI from '@/types/ui';

import { ActionClickHandler } from '@/decorators/ActionHandlerDecorator';
import { Trans } from 'react-i18next';
import { parseTranslationModules } from '@/utils/TranslationUtils';

const getDivider = (menuIndex: number, itemIndex: number) => {
  return {
    id: `divider${menuIndex}_${itemIndex}`,
  };
};

const actionToActionMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  action: UI.MenuActionDefition<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
) => {
  const active = !action.checked ? false : action.checked(menu.itemDataGetter());
  const icon = !!action.checked ? (active ? 'checkmark' : '') : action.icon;
  return {
    id: action.id,
    item: {
      onClick: () => {
        onClickAction({
          action,
          itemData: menu.itemDataGetter(),
          entity: menu.entity,
          moduleData: menu.moduleData,
        });
      },
      active,
      icon,
      children: (
        <Trans
          i18nKey={toActionI18nKey(
            action,
            parseTranslationModules(menu.moduleData.moduleId),
          )}
          defaults={action.displayName}
        >
          {action.displayName}
        </Trans>
      ),
    },
  };
};

// Convert ID to menu link element
const getMenuItem = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
  action: UI.MenuActionListItemType<ItemDataT, EntityT>,
  itemIndex: number,
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
): UI.ActionMenuItem => {
  if (!action) {
    return getDivider(menuIndex, itemIndex);
  }

  if ('children' in action) {
    const children = action.children.map((childAction, childActionIndex) => {
      if (!childAction) {
        return getDivider(0, childActionIndex);
      }

      return actionToActionMenuItem(childAction, menu, onClickAction);
    });

    if (children.length === 1) {
      return children[0];
    }

    return {
      id: action.id,
      item: {
        icon: action.icon,
        onClick: () => {
          // ..
        },
        children: <>{action.displayName}</>,
      },
      children,
    };
  }

  return actionToActionMenuItem(action, menu, onClickAction);
};

export const localMenuToActionMenuItems = <
  ItemDataT extends UI.ActionMenuItemDataValueType,
  EntityT extends UI.ActionMenuItemEntityValueType,
>(
  onClickAction: ActionClickHandler<ItemDataT, EntityT>,
  menu: UI.ActionMenuType<ItemDataT, EntityT>,
  menuIndex: number,
) => {
  const items = menu.actions.map((actionId, actionIndex) => {
    return getMenuItem(
      menu as UI.ActionMenuType<ItemDataT, EntityT>,
      menuIndex,
      actionId,
      actionIndex,
      onClickAction,
    );
  });

  return items;
};
