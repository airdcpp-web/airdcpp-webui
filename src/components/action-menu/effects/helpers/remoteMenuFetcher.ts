import { useEffect, useState } from 'react';

import * as API from 'types/api';
import * as UI from 'types/ui';

import MenuConstants, { MENU_SUPPORTS } from 'constants/MenuConstants';
import NotificationActions from 'actions/NotificationActions';
import { parseActionMenuItemId, parseActionMenuItemIds } from 'utils/MenuUtils';
import { ActionMenuDefinition } from '../useActionMenuItems';
import { useSocket } from 'context/SocketContext';

export interface RemoteMenuData {
  selectedIds: UI.ActionIdType[];
  remoteMenuId: string;
  entityId: API.IdType | undefined;
}

export interface RemoteMenu {
  menuData: RemoteMenuData;
  items: API.GroupedContextMenuItem[];
}

interface ParsedDefinitionArray {
  remoteMenuIds: Array<string | undefined>;
  selectedIdsByMenu: Array<UI.ActionIdType>[];
  entityIds: Array<API.IdType | undefined>;
}

const parseDefinitionArray = (
  definitionArray: ActionMenuDefinition<any, any>[],
): ParsedDefinitionArray => {
  const selectedIdsByMenu = definitionArray
    .map((defition) => defition.itemData)
    .map((data) => parseActionMenuItemIds(data));
  const remoteMenuIds = definitionArray.map(
    (remoteMenuProps) => remoteMenuProps.remoteMenuId,
  );
  const entityIds = definitionArray
    .map((defition) => defition.entity)
    .map((data) => parseActionMenuItemId(data) as API.IdType | undefined);

  return { selectedIdsByMenu, remoteMenuIds, entityIds };
};

export const useRemoteMenuFetcher = (
  definitionArray: ActionMenuDefinition<any, any>[],
) => {
  const { selectedIdsByMenu, remoteMenuIds, entityIds } =
    parseDefinitionArray(definitionArray);

  const [menusByType, setMenusByType] = useState<
    Array<RemoteMenu | undefined> | undefined
  >(undefined);

  const socket = useSocket();

  useEffect(() => {
    const fetchMenus = async () => {
      const menuPromises = remoteMenuIds.map(async (remoteMenuId, menuIndex) => {
        if (!remoteMenuId) {
          return Promise.resolve(undefined);
        }

        const selectedIds = selectedIdsByMenu[menuIndex];
        const entityId = entityIds[menuIndex];

        const menu = await socket.post<API.GroupedContextMenuItem[]>(
          `${MenuConstants.MODULE_URL}/${remoteMenuId}/list_grouped`,
          {
            selected_ids: selectedIds,
            entity_id: entityId,
            supports: MENU_SUPPORTS,
          },
        );

        return {
          items: menu,
          menuData: {
            remoteMenuId,
            selectedIds,
            entityId,
          },
        };
      });

      let fetchedMenus: Array<RemoteMenu | undefined>;
      try {
        fetchedMenus = await Promise.all(menuPromises);
      } catch (e) {
        NotificationActions.apiError('Failed fetch menu items', e);
        return;
      }

      setMenusByType(fetchedMenus);
    };

    fetchMenus();
  }, []);

  return menusByType;
};
