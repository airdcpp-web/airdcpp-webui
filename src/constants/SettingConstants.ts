const MODULE_URL = 'settings';

export const enum SettingProfileEnum {
  NORMAL = 0,
  RAR = 1,
  LAN = 2,
}

export const enum IncomingConnectionModeEnum {
  INCOMING_DISABLED = -1,
  INCOMING_ACTIVE = 0,
  INCOMING_ACTIVE_UPNP = 1,
  INCOMING_PASSIVE = 2,
}

export const enum OutgoingConnectionModeEnum {
  OUTGOING_DIRECT = 0,
  OUTGOING_SOCKS = 1,
}

export const enum LocalSettings {
  NOTIFY_MENTION = 'notify_mention_my_nick',
  NOTIFY_PM_USER = 'notify_pm_user',
  NOTIFY_PM_BOT = 'notify_pm_bot',
  NOTIFY_HUB_MESSAGE = 'notify_hub_message',

  NOTIFY_BUNDLE_STATUS = 'notify_bundle_status',

  NOTIFY_EVENTS_INFO = 'notify_events_info',
  NOTIFY_EVENTS_WARNING = 'notify_events_warning',
  NOTIFY_EVENTS_ERROR = 'notify_events_error',

  UNREAD_LABEL_DELAY = 'unread_label_delay',
  BACKGROUND_IMAGE_URL = 'background_image_url',

  NO_INSTALL_PROMPT = 'no_install_prompt',
}

export default {
  MODULE_URL: MODULE_URL,

  ITEMS_GET_URL: MODULE_URL + '/get',
  ITEMS_SET_URL: MODULE_URL + '/set',
  ITEMS_RESET_URL: MODULE_URL + '/reset',
  ITEMS_DEFINITIONS_URL: MODULE_URL + '/definitions',
};
