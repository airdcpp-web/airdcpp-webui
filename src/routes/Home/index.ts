import IconConstants from 'constants/IconConstants';

import * as UI from 'types/ui';

export const HOME_URL = '/home';

export const HomeRoutes: UI.RouteItem = {
  title: 'Home',
  path: HOME_URL,
  icon: IconConstants.HOME,
  lazy: async () => {
    const { default: Component } = await import(
      /* webpackChunkName: "home" */ './components'
    );
    return {
      Component,
    };
  },
};
