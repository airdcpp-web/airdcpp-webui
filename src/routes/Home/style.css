/* Layout */
#home .ui.message .content .description {
  margin-top: 15px;
}

#home .ui.message .button {
  margin-top: 10px;
}

#mobile-layout #home {
  height: 100%;
  overflow-y: auto;
  overflow-x: hidden;
}

#mobile-layout #home > div {
  margin-left: 8px !important;
  margin-right: 8px !important;
}

#mobile-layout #home .ui.message {
  margin-top: 10px;
}

#mobile-layout .react-grid-layout {
  margin-top: 0px;
  margin-bottom: 25px;
}

/* Widgets */
.react-grid-layout .widget .main.content {
  max-height: calc(100% - 45px);
  display: flex;
  overflow: hidden;
}

.react-grid-layout .widget .main.content > div {
  max-height: 100%;
  max-width: 100%;
  display: flex;
  width: 100%;
  flex: 1;
}

.react-grid-layout .widget .header i.large.icon {
  font-size: 1.3em !important;
}

.react-grid-layout .widget .header-row {
  display: flex;
  flex: 0 0 auto !important;
}

.react-grid-layout .widget .header-row .header {
  flex: 1;
  height: 100%;
}

.react-grid-layout .widget .main.content .extra.content {
  opacity: 0.4;
}

#mobile-layout .react-grid-layout .react-grid-item {
  /* Allow touch scrolling */
  touch-action: initial !important;
  margin-left: 0px !important;
  margin-right: 0px !important;
}
