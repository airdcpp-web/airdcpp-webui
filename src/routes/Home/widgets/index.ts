import { ApplicationWidgetInfo } from './Application';
import { ExtensionWidgetInfo } from './Extensions';
import { NotepadWidgetInfo } from './Notepad';
import { RSSWidgetInfo } from './RSS';
import { TransferWidgetInfo } from './Transfers';

export const Widgets = [
  ApplicationWidgetInfo,
  RSSWidgetInfo,
  NotepadWidgetInfo,
  TransferWidgetInfo,
  ExtensionWidgetInfo,
];

export * from './Application';
export * from './Extensions';
export * from './Notepad';
export * from './RSS';
export * from './Transfers';
