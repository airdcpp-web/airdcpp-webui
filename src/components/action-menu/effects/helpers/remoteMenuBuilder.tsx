import * as API from '@/types/api';
import * as UI from '@/types/ui';

import MenuConstants from '@/constants/MenuConstants';
import IconConstants from '@/constants/IconConstants';
import { RemoteMenu, RemoteMenuData } from './remoteMenuFetcher';
import { UILogger } from '@/effects/LoggerEffect';

export interface CombinedContextMenuItem {
  typeData: RemoteMenuData;
  item: API.ContextMenuItem;
}

export interface CombinedContextMenu extends Omit<API.GroupedContextMenu, 'items'> {
  items: CombinedContextMenuItem[];
}

interface MenuOptions {
  handleSelect: SelectHandler;
  nestingThreshold: number;
}

type SelectHandler = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
) => void;

const toMenuItem = (
  menuItem: API.ContextMenuItem,
  remoteMenuData: RemoteMenuData,
  handleSelect: SelectHandler,
): UI.ActionMenuItem => {
  const onClick = () => {
    handleSelect(menuItem, remoteMenuData);
  };

  const { id, icon, title } = menuItem;
  return {
    id,
    item: {
      onClick,
      icon: icon[MenuConstants.ICON_TYPE_ID],
      children: title,
    },
  };
};

export const remoteMenuToActionMenuItems = (
  menu: CombinedContextMenu,
  { handleSelect, nestingThreshold }: MenuOptions,
): UI.ActionMenuItem[] => {
  const items = menu.items.map(({ item, typeData }) =>
    toMenuItem(item, typeData, handleSelect),
  );

  if (items.length <= nestingThreshold) {
    return items;
  }

  return [
    {
      id: menu.id,
      item: {
        icon: menu.icon[MenuConstants.ICON_TYPE_ID] || IconConstants.EXTENSION,
        onClick: () => {
          // ..
        },
        children: <>{menu.title}</>,
      },
      children: items,
    },
  ];
};

// Combine menus by their subscriber ID and remove duplicate items
export const createRemoteMenuCombineReducer = (logger: UILogger) => {
  return (reduced: CombinedContextMenu[], typeMenus: RemoteMenu | undefined) => {
    if (!typeMenus) {
      return reduced;
    }

    for (const menu of typeMenus.menus) {
      const existing = reduced.find((m) => m.id === menu.id);

      const toMenuItem = (item: API.ContextMenuItem): CombinedContextMenuItem => ({
        typeData: typeMenus.typeData,
        item,
      });

      if (existing) {
        existing.items.push(
          ...menu.items
            .filter((item) => {
              const existingItem = existing.items.find(
                (existingItem) =>
                  existingItem.item.id === item.id &&
                  existingItem.item.hook_id === item.hook_id,
              );
              if (existingItem) {
                logger.info(
                  // eslint-disable-next-line max-len
                  `Removing a duplicate item ${item.id} (subscriber ${menu.id}) from menu ${typeMenus.typeData.remoteMenuId} (exists for menu ${existingItem.typeData.remoteMenuId})`,
                );
              }

              return !existingItem;
            })
            .map(toMenuItem),
        );
      } else {
        reduced.push({
          ...menu,
          items: menu.items.map(toMenuItem),
        });
      }
    }

    return reduced;
  };
};
