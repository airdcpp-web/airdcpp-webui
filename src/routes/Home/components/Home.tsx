import { Component } from 'react';

import DemoIntro from '@/routes/Home/components/DemoIntro';
import InstallPrompt from '@/components/InstallPrompt';
import NewUserIntro from '@/routes/Home/components/NewUserIntro';
import WidgetLayout from '@/routes/Home/components/WidgetLayout';

import '../style.css';

class Home extends Component {
  static readonly displayName = 'Home';

  render() {
    return (
      <div id="home">
        <DemoIntro />
        <NewUserIntro />
        <InstallPrompt />
        <WidgetLayout {...this.props} />
      </div>
    );
  }
}

export default Home;
