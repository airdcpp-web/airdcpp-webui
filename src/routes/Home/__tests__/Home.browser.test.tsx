import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import Home from '../components/Home';
import { getMockSession } from '@/tests/mocks/mock-session';
import SettingConstants from '@/constants/SettingConstants';
import { clickButton, expectResponseToMatchSnapshot } from '@/tests/helpers/test-helpers';

import createFetchMock from 'vitest-fetch-mock';
import { ReleaseFeedRSSContent } from '@/tests/mocks/http/rss-feed-content';
import { ExtensionReleasesContent } from '@/tests/mocks/http/extension-releases-content';
import {
  installApplicationWidgetMocks,
  installExtensionWidgetMocks,
  installTransferWidgetMocks,
} from '@/tests/mocks/mock-widget';
import { waitFor } from '@testing-library/dom';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';

const fetchMocker = createFetchMock(vi);

describe('Home layout', () => {
  let server: MockServer;
  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);
    commonData.appStore.getState().login.onLoginCompleted(
      commonData.socket,
      {
        ...getMockSession(),
        wizard_pending: true,
      },
      true,
    );

    installTransferWidgetMocks(server);
    installApplicationWidgetMocks(server);
    installExtensionWidgetMocks(server);

    return { commonData, server };
  };

  const renderLayout = async (settings: object = {}) => {
    const { commonData, server, ...other } = await getSocket();

    const HomeLayoutTest = () => {
      return <Home />;
    };

    const routes = [
      {
        path: '/home/*',
        Component: HomeLayoutTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/home'] },
    });

    return { ...renderData, ...other, ...commonData };
  };

  beforeEach(() => {
    server = getMockServer();

    fetchMocker.enableMocks();

    fetchMocker.mockIf(/.*$/, (req) => {
      if (req.url.includes('feed.xml')) {
        return ReleaseFeedRSSContent;
      } else if (req.url.includes('airdcpp-extensions-public')) {
        return ExtensionReleasesContent;
      } else {
        return {
          status: 404,
          body: 'Not Found',
        };
      }
    });
  });

  afterEach(() => {
    server.stop();
  });

  test('should handle new user intro', async () => {
    const { getByRole, queryByRole } = await renderLayout();

    const onIntroSeen = vi.fn();
    server.addRequestHandler(
      'POST',
      SettingConstants.ITEMS_SET_URL,
      undefined,
      onIntroSeen,
    );

    const wizardCloseButtonCaption = "Close and don't show again";
    clickButton(wizardCloseButtonCaption, getByRole);

    await waitFor(() => {
      expect(onIntroSeen).toHaveBeenCalledTimes(1);
    });

    expectResponseToMatchSnapshot(onIntroSeen);

    await waitFor(() =>
      expect(
        queryByRole('button', { name: wizardCloseButtonCaption }),
      ).not.toBeInTheDocument(),
    );
  });
});
