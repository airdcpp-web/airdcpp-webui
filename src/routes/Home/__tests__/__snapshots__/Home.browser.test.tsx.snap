// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home layout > should handle new user intro 1`] = `
{
  "data": {
    "wizard_pending": false,
  },
  "method": "POST",
  "path": "settings/set",
}
`;
